### PR TITLE
Add data types for Slash Commands

### PR DESCRIPTION
**Pull request by UnderMybrella**
Adds all the new slash command types from https://discord.com/developers/docs/interactions/slash-commands#data-models-and-types

### CODE REVIEW

File: src/main/java/discord4j/discordjson/json/ApplicationCommandOptionData.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    String description();
+
+    /**
+     * the first required option for the user to complete -- only one option can be default
+     */
+    @JsonProperty("default")
+    Possible<Boolean> isDefault();
+
+    /**
+     * if the fparameter is required or optional -- default false
```
Comment (Shadorc) [inline src/main/java/discord4j/discordjson/json/ApplicationCommandOptionData.java:42]:
```suggestion
     * if the parameter is required or optional -- default false
```

File: src/main/java/discord4j/discordjson/json/ApplicationCommandOptionData.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+     * 1-100 character description
+     */
+    String description();
+
+    /**
+     * the first required option for the user to complete -- only one option can be default
+     */
+    @JsonProperty("default")
+    Possible<Boolean> isDefault();
```
Comment (Shadorc) [inline src/main/java/discord4j/discordjson/json/ApplicationCommandOptionData.java:39]: It should be `default()` to be consistent with the rest of the API. For example, just below, `required()`

File: src/main/java/discord4j/discordjson/json/ApplicationCommandInteractionOptionData.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    /**
+     * the name of the parameter
+     */
+    String name();
+
+    /**
+     * the value of the pair
+     */
+    Possible<String> value();
```
Comment (Shadorc) [inline src/main/java/discord4j/discordjson/json/ApplicationCommandInteractionOptionData.java:27]: Isn't it an int? https://discord.com/developers/docs/interactions/slash-commands#applicationcommandoptiontype

**Review by Shadorc — CHANGES_REQUESTED** [review]

File: src/main/java/discord4j/discordjson/json/ApplicationCommandOptionData.java (hunk 2 above)
Comment (Shadorc) [reply src/main/java/discord4j/discordjson/json/ApplicationCommandOptionData.java:39]: Oh, nevermind, I forgot default is a reserved name

File: src/main/java/discord4j/discordjson/json/ApplicationCommandInteractionOptionData.java (hunk 3 above)
Comment (UnderMybrella) [reply src/main/java/discord4j/discordjson/json/ApplicationCommandInteractionOptionData.java:27]: Not quite; value is of *type* OptionType, which allows receiving a value of type user, channel, string, etc. I've made this a String here as it's a supertype of all the serialised values.

Comment (Shadorc) [reply src/main/java/discord4j/discordjson/json/ApplicationCommandInteractionOptionData.java:27]: Oh okay, so what are the values (int) in the doc I linked?

Comment (Shadorc) [reply src/main/java/discord4j/discordjson/json/ApplicationCommandInteractionOptionData.java:27]: Thanks for the explanation, I now understand

File: src/main/java/discord4j/discordjson/json/InteractionData.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import discord4j.discordjson.json.gateway.Dispatch;
+import discord4j.discordjson.possible.Possible;
+import org.immutables.value.Value;
+
+@Value.Immutable
+@JsonSerialize(as = ImmutableInteractionData.class)
+@JsonDeserialize(as = ImmutableInteractionData.class)
+public interface InteractionData extends Dispatch {
```
Comment (quanticc) [inline src/main/java/discord4j/discordjson/json/InteractionData.java:13]: As discussed, we need to split InteractionCreate from InteractionData.

The former can contain the latter using JsonUnwrapped.

**Review by quanticc — CHANGES_REQUESTED** [review]
Thanks for the PR. need some changes before merging

File: src/main/java/discord4j/discordjson/json/InteractionData.java (hunk 4 above)
Comment (UnderMybrella) [reply src/main/java/discord4j/discordjson/json/InteractionData.java]: I somehow didn't push my changes, that's on me. That should fix this.

**Review by quanticc — APPROVED** [review]

**Comment by darichey** [comment]
Correct me if I'm wrong, but I think we missed a `ModifyApplicationCommandRequest` here. It seems like `ApplicationCommandRequest` was used for both the create and edit routes, but this is not correct, because all fields of the latter should be optional.

**Pull request merged**